app/permissions_migrations.go

```diff
@@ -0,0 +1,157 @@
+		for _, f := range funcs {
+			if f(permissions) {
+				return true
+			}
+		}
+		return false
+	}
+}
+
+func permissionAnd(funcs ...func(map[string]bool) bool) func(map[string]bool) bool {
+	return func(permissions map[string]bool) bool {
+		for _, f := range funcs {
+			if !f(permissions) {
+				return false
+			}
+		}
+		return true
+	}
+}
+
+func applyPermissionsMap(permissions []string, migrationMap permissionsMap) []string {
+	finalMap := make(map[string]bool)
+	var result []string
+	for _, permission := range permissions {
+		finalMap[permission] = true
+	}
+
+	for _, transformation := range migrationMap {
+		if transformation.On(finalMap) {
+			for _, add := range transformation.Add {
+				finalMap[add] = true
+			}
+			for _, remove := range transformation.Remove {
+				finalMap[remove] = false
+			}
+		}
+	}
+
+	for key, active := range finalMap {
+		if active {
+			result = append(result, key)
+		}
+	}
+	return result
+}
+
+func (a *App) doPermissionsMigration(key string, migrationMap permissionsMap) *model.AppError {
+	if result := <-a.Srv.Store.System().GetByName(key); result.Err == nil {
+		return nil
+	}
+
+	roles, err := a.GetRoles()
+	if err != nil {
+		return err
+	}
+
+	for _, role := range roles {
+		role.Permissions = applyPermissionsMap(role.Permissions, migrationMap)
+		if result := <-a.Srv.Store.Role().Save(role); result.Err != nil {
+			return result.Err
+		}
+	}
+
+	if result := <-a.Srv.Store.System().Save(&model.System{Name: key, Value: "true"}); result.Err != nil {
+		return result.Err
+	}
+	return nil
+}
+
+func getEmojisPermissionsSplitMigration() permissionsMap {
+	return permissionsMap{
+		permissionTransformation{
+			On:     permissionExists(PERMISSION_MANAGE_EMOJIS),
+			Add:    []string{PERMISSION_CREATE_EMOJIS, PERMISSION_DELETE_EMOJIS},
+			Remove: []string{PERMISSION_MANAGE_EMOJIS},
+		},
+		permissionTransformation{
+			On:     permissionExists(PERMISSION_MANAGE_OTHERS_EMOJIS),
+			Add:    []string{PERMISSION_DELETE_OTHERS_EMOJIS},
+			Remove: []string{PERMISSION_MANAGE_OTHERS_EMOJIS},
+		},
+	}
+}
+
+func getWebhooksPermissionsSplitMigration() permissionsMap {
+	return permissionsMap{
+		permissionTransformation{
+			On:     permissionExists(PERMISSION_MANAGE_WEBHOOKS),
+			Add:    []string{PERMISSION_MANAGE_INCOMING_WEBHOOKS, PERMISSION_MANAGE_OUTGOING_WEBHOOKS},
+			Remove: []string{PERMISSION_MANAGE_WEBHOOKS},
+		},
+		permissionTransformation{
+			On:     permissionExists(PERMISSION_MANAGE_OTHERS_WEBHOOKS),
+			Add:    []string{PERMISSION_MANAGE_OTHERS_INCOMING_WEBHOOKS, PERMISSION_MANAGE_OTHERS_OUTGOING_WEBHOOKS},
+			Remove: []string{PERMISSION_MANAGE_OTHERS_WEBHOOKS},
+		},
+	}
+}
+
+// DoPermissionsMigrations execute all the permissions migrations need by the current version.
+func (a *App) DoPermissionsMigrations() *model.AppError {
+	if err := a.doPermissionsMigration(MIGRATION_KEY_EMOJI_PERMISSIONS_SPLIT, getEmojisPermissionsSplitMigration()); err != nil {
```

Comment:
Since there's going to be many more of these, how about a (statically declared) ordered list of the key/function pairs, and this function just looping over that executing them and failing on the errors? (0/5)

Reply:
I like it.

Reply:
Done